daisy/workflow/sources.go

```diff
@@ -0,0 +1,118 @@
+package workflow
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+)
+
+func (w *Workflow) recursiveGCS(bkt, prefix, dst string) error {
+	it := w.StorageClient.Bucket(bkt).Objects(w.Ctx, &storage.Query{Prefix: prefix})
+	for {
+		objAttr, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if objAttr.Size == 0 {
+			continue
+		}
+		srcPath := w.StorageClient.Bucket(bkt).Object(objAttr.Name)
+		o := path.Join(w.sourcesPath, dst, strings.TrimPrefix(objAttr.Name, prefix))
+		dstPath := w.StorageClient.Bucket(w.bucket).Object(o)
+		if _, err := dstPath.CopierFrom(srcPath).Run(w.Ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *Workflow) uploadFile(src, obj string) error {
+	dstPath := w.StorageClient.Bucket(w.bucket).Object(path.Join(w.sourcesPath, obj))
+	gcs := dstPath.NewWriter(w.Ctx)
+	f, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(gcs, f); err != nil {
+		return err
+	}
+	return gcs.Close()
+}
+
+func (w *Workflow) uploadSources() error {
+	for dst, origPath := range w.Sources {
+		// GCS to GCS.
+		if bkt, prefix, err := splitGCSPath(origPath); err == nil {
+			if prefix == "" {
+				if err := w.recursiveGCS(bkt, prefix, dst); err != nil {
+					return err
+				}
+				continue
+			}
+			src := w.StorageClient.Bucket(bkt).Object(path.Clean(prefix))
+			// If this is a GCS 'directory' we will get ErrObjectNotExist.
```

Comment:
Yep, fixed so in this case the user would have to have the gcs path as gs://bucket/object/